pkg/api/types_test.go

```diff
@@ -292,152 +296,122 @@ func TestJSONTags(t *testing.T) {
+	// TODO: why don't we just include all of these in the admin type?
+	notInAdmin := []*regexp.Regexp{
+		regexp.MustCompile(`^\.Config\.RunningUnderTest$`),
+		regexp.MustCompile(`^\.Properties\.AzProfile\.`),
+		regexp.MustCompile(`^\.Properties\.ServicePrincipalProfile\.`),
 	}
 
-	adminIndexOffset := 0
-	internalIndexOffeset := 0
-	for i := 0; i < internalNum; i++ {
-		parentPrefix := parent + "."
-		internalTag := filterName(internalType.Field(i - internalIndexOffeset).Tag)
-		if _, ok := internalIgnoredFields[parent][internalTag]; ok {
-			t.Logf("ignoring internal field %q", parentPrefix+internalTag)
-			adminIndexOffset++
-			continue
-		}
-		adminTag := filterName(adminType.Field(i - adminIndexOffset).Tag)
-		if _, ok := adminIgnoredFields[parent][adminTag]; ok {
-			t.Logf("ignoring admin field %q", parentPrefix+adminTag)
-			internalIndexOffeset++
-			continue
+loop:
+	for len(afields) > 0 || len(ifields) > 0 {
+		if len(afields) > 0 {
+			for _, rx := range notInInternal {
+				if rx.MatchString(afields[0]) {
+					afields = afields[1:]
+					continue loop
+				}
+			}
 		}
 
-		t.Logf("comparing internal field tag %q with admin field tag %q", parentPrefix+internalTag, parentPrefix+adminTag)
-		if internalTag != adminTag {
-			t.Fatalf("admin field tag is %q, internal field tag is %q", adminTag, internalTag)
-		}
+		if len(ifields) > 0 {
+			if strings.Contains(ifields[0], "Key") ||
+				strings.Contains(ifields[0], "Kubeconfig") ||
+				strings.Contains(ifields[0], "Passwd") ||
+				strings.Contains(ifields[0], "Password") ||
+				strings.Contains(ifields[0], "Secret") {
+				ifields = ifields[1:]
+				continue
+			}
 
-		internalFieldValue := internalValue.Field(i - internalIndexOffeset)
-		if internalFieldValue.Kind() == reflect.Ptr {
-			internalFieldValue = reflect.Indirect(internalFieldValue)
-		}
-		adminFieldValue := adminValue.Field(i - adminIndexOffset)
-		if adminFieldValue.Kind() == reflect.Ptr {
-			adminFieldValue = reflect.Indirect(adminFieldValue)
+			for _, rx := range notInAdmin {
+				if rx.MatchString(ifields[0]) {
+					ifields = ifields[1:]
+					continue loop
+				}
+			}
 		}
 
-		if internalFieldValue.Type().Name() == "Certificate" {
-			// Ignore comparing x509.Certificate
+		if len(afields) > 0 && len(ifields) > 0 && afields[0] == ifields[0] {
+			afields, ifields = afields[1:], ifields[1:]
 			continue
 		}
 
-		if internalFieldValue.Kind() == reflect.Interface {
-			internalInterface := internalFieldValue.Interface()
-			if internalAad, ok := internalInterface.(*AADIdentityProvider); ok {
-				internalFieldValue = reflect.ValueOf(*internalAad)
-			}
-			adminInterface := adminFieldValue.Interface()
-			if adminAad, ok := adminInterface.(*admin.AADIdentityProvider); ok {
-				adminFieldValue = reflect.ValueOf(*adminAad)
-			}
+		if len(afields) > 0 {
+			t.Error(afields[0])
```

Comment:
The error produced here is very unclear, you should at least print whether it's an internal or external field. Likewise in L413.

Reply:
Agreed - updated